rids/rules.py

```diff
@@ -0,0 +1,103 @@
+# Copyright 2022 Jigsaw Operations LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Generalizes the various formats of IOCs into Rule and RuleSet.
+
+Rule is a value type that represents evaluation on a single rule.  RuleSet
+instances represent a set of independent and overlapping Rule instances.
+
+RuleSet should typically be constructed from one of the flavors of parsers
+in `rids.iocs` or merged from combining a pair of RuleSet instances.  Its
+
+Evaluation of the RuleSet is done with the ProcessHandshake(...) and
+ProcessEndpoint(...) methods which take an observation and return a list
+of zero or more Event instances with their matching rules in the RuleSet.
+"""
+
+
+import collections
+
+
+class RuleSet:
+    __slots__ = ('ip_address_rules', )
+
+    def __init__(self):
+      self.ip_address_rules = collections.defaultdict(list)
+
+    def MergeRuleset(self, other):
+      """Simple implementation of a merging of rule sets."""
+      for key, value in other.ip_address_rules.items():
+        self.ip_address_rules[key].extend(value)
+
+    def AddRule(self, rule):
+      """Simple implementation of adding a single rule to this rule set."""
+      if hasattr(rule, 'matches_ip'):
+        self.ip_address_rules[str(rule.matches_ip)].append(rule)
+
+    def ProcessHandshake(self, observation):
+      """Process a handshake-related observation."""
+      return Event(observation)
+
+    def ProcessEndpoint(self, observation):
+      """Process observations related to a remote IP address."""
+      ip_str = str(observation['remote_ip'])
+      events = []
+      if ip_str in self.ip_address_rules:
+        for rule in self.ip_address_rules[ip_str]:
+          event = {}
+          event.update(observation)
+          event.update(rule.as_dict())
+          events.append(event)
+      return events
+
+
+class Rule:
```

Comment:
No need for this generic Rule for now, which is not well defined. Create a `IpRule` instead, since that's what we have.

Reply:
I know from reading a lot of suricata rules, especially the ones that are fundamentally about disallowing IPs, the kind of other metadata they typically contain (msg, reference, sometimes a creation timestamp, sometimes the class of services they talk to is important in the blocking decision) and I was starting this Rule definition based on these rules, so it's not as poorly defined as you make it sound.  But it is very loosely typed right now, I agree.

I think in the long term we'll benefit from a single generally-defined Rule type that gets indexed appropriately when being added to RuleSet, but we can try individual and specific rule types first.  A lot of rules will share a subset of properties, mostly for communicating the rule or for rate-dependent detection, but even the suricata schema leaves all of them as effectively optional fields (except action and the header describing the endpoints to match against, those are always defined).

Reply:
done